### DISCUSSION TITLE
[dnsdist] How to execute SpoofAction in LuaResponseAction

### DISCUSSION THREAD

**Discussion by KagurazakaNyaa** [PowerDNS/pdns#14795] [Q&A] [answered]
Now I try to add a tag to `dr` in `LuaResponseAction` and use `dr:restart()` to re-trigger the query, but it doesn't seem to route the request correctly based on the tag.

## config

```lua
local function readCloudflareFastIPs()
    local cloudflare_fast = ReadFile("/etc/dnsdist/data/cloudflare_fast.txt")
    return cloudflare_fast
end

function makeQueryRestartable(dq)
    -- make it possible to restart that query later
    -- by keeping a copy of the initial DNS payload around
    dq:setRestartable()
    return DNSAction.None
end
-- if response has cloudflare ip ,tag it and restart
local function tagResponseCloudflare(dr)
    if dr.pool == 'local' then
        return DNSResponseAction.None
    end
    local packet = dr:getContent()
    if string.len(packet) == 0 then
        return DNSResponseAction.Drop
    end
    local overlay = newDNSPacketOverlay(packet)
    local count = overlay:getRecordsCountInSection(DNSSection.Answer)
    for i = 0, count - 1 do
        local record = overlay:getRecord(i)
        local parser = parsers[recordTypes[record.type]]
        if parser then
            local ca = parser(packet, record.contentOffset + 1, record.contentLength)
            if cloudflare_netmask_group:match(ca) then
                print("cloudflare IP: " .. ca:tostring() .. ", need replace")
                dr:setTag("cloudflare_fast", "need_replace")
                dr:restart()
                print("dropping resp")
                return DNSResponseAction.Drop
            end
        end
    end
    return DNSResponseAction.None
end

local cloudflare_fast_bypass = SetTagAction("cloudflare_fast", "bypass")
local cloudflare_fast_replace = TagRule("cloudflare_fast", "need_replace")
local fast_cloudflare_action = SpoofAction(readCloudflareFastIPs())
local cloudflare_fast_qtype = OrRule({ QTypeRule(DNSQType.A), QTypeRule(DNSQType.AAAA) })
local cloudflare_fast_not_optimized = NotRule(TagRule("cloudflare_fast"))
local cloudflare_fast_need_optimize = AndRule({ cloudflare_fast_not_optimized, cloudflare_fast_qtype })

-- if dq is set cloudflare_fast, spoof
addAction(cloudflare_fast_replace, fast_cloudflare_action, { name = "cloudflare_fast_spoof" })

-- unmatch china domains and proxy domains request will use local pool first and proecss in response actions
addAction(AllRule(), LuaAction(makeQueryRestartable), { name = "make_query_restartable" })
addAction(AllRule(), PoolAction("local"), { name = "default_use_local" })

-- post process
addResponseAction(cloudflare_fast_need_optimize, LuaResponseAction(tagResponseCloudflare),
    { name = "tag_cloudflare_resp_fast" })
```
cloudflare_fast.txt
```
104.31.221.176
104.31.230.80
104.31.147.21
104.31.223.195
104.31.97.33
```

## test
I tried to execute the request `nslookup chat.waw-eve.com 127.0.0.1`, and it returned
```
Server:		127.0.0.1
Address:	127.0.0.1#53


Non-authoritative answer:
Name:	chat.waw-eve.com
Address: 104.21.70.120
Name:	chat.waw-eve.com
Address: 172.67.223.100
Name:	chat.waw-eve.com
Address: 2606:4700:3030::ac43:df64
Name:	chat.waw-eve.com
Address: 2606:4700:3030::6815:4678
```
The log shows the following
```log
cloudflare IP: 104.21.70.120, need replace
dropping resp
cloudflare IP: 2606:4700:3030::6815:4678, need replace
```
Therefore, it can be confirmed that the request is correctly routed to `tagResponseCloudflare` and `dr:restart()` is triggered.

However, the returned result is not replaced, and I see on the webui that `addAction(cloudflare_fast_replace, fast_cloudflare_action, { name = "cloudflare_fast_spoof" })` is not triggered

![image](https://github.com/user-attachments/assets/de47c68e-d0e2-460a-a974-ce1989143e55)

**Comment by rgacogne** [comment]
What is happening is that you seem to expect the "query" rules (the ones added via `addAction`) to be executed for a restarted query, which is not the case because most of the time it is not what the user wants. Cache miss rules are executed, if that helps.
I'm not sure I understand what your actual needs are, and I cannot really test the configuration you shared here anyway because some parts are missing (`ReadFile`, `parsers` at least).

**Reply by KagurazakaNyaa** [reply]
> What is happening is that you seem to expect the "query" rules (the ones added via `addAction`) to be executed for a restarted query, which is not the case because most of the time it is not what the user wants. Cache miss rules are executed, if that helps. I'm not sure I understand what your actual needs are, and I cannot really test the configuration you shared here anyway because some parts are missing (`ReadFile`, `parsers` at least).

My intention is to try to tag and re-execute the request based on the dnsresponse content obtained from the upstream and have this request routed to a `SpoofAction` based on this tag.
This behavior is expected to cause all DNS requests that return IPs belonging to Cloudflare to be modified to a few fixed IPs.

```lua
-- see if the file exists
function CheckFileExists(file)
    local f = io.open(file, "rb")
    if f then f:close() end
    return f ~= nil
end

-- get all lines from a file, returns an empty
-- list/table if the file does not exist
function ReadFile(file)
    if not CheckFileExists(file) then return {} end
    local lines = {}
    for line in io.lines(file) do
        lines[#lines + 1] = line
    end
    return lines
end
```

```lua
-- rules define
local parsers = {};
local recordTypes = { [1] = 'A', [28] = 'AAAA' };

function parsers.A(packet, pos, contentLength)
    if contentLength == 4 then
        return newCA(table.concat({ packet:byte(pos, pos + 3) }, "."));
    end
end

function parsers.AAAA(packet, pos, contentLength)
    if contentLength == 16 then
        local t = { packet:byte(pos, pos + 15) };
        for i = 1, 8 do
            t[i] = ("%x"):format(t[i * 2 - 1] * 256 + t[i * 2]);
        end
        return newCA(table.concat(t, ":", 1, 8));
    end
end

-- match cloudflare ips
local cloudflare_ipv4 = {
    "173.245.48.0/20",
    "103.21.244.0/22",
    "103.22.200.0/22",
    "103.31.4.0/22",
    "141.101.64.0/18",
    "108.162.192.0/18",
    "190.93.240.0/20",
    "188.114.96.0/20",
    "197.234.240.0/22",
    "198.41.128.0/17",
    "162.158.0.0/15",
    "104.16.0.0/13",
    "104.24.0.0/14",
    "172.64.0.0/13",
    "131.0.72.0/22",
}
local cloudflare_ipv6 = {
    "2400:cb00::/32",
    "2606:4700::/32",
    "2803:f800::/32",
    "2405:b500::/32",
    "2405:8100::/32",
    "2a06:98c0::/29",
    "2c0f:f248::/32",
}
local cloudflare_netmask_group = newNMG();
for _, ip in ipairs(cloudflare_ipv4) do
    cloudflare_netmask_group:addMask(ip)
end
for _, ip in ipairs(cloudflare_ipv6) do
    cloudflare_netmask_group:addMask(ip)
end
```

So should I add this rule in `addCacheInsertedResponseAction`?

**Reply by rgacogne** [reply]
> My intention is to try to tag and re-execute the request based on the dnsresponse content obtained from the upstream and have this request routed to a SpoofAction based on this tag.

Right, then I'm not sure you actually need to restart a query, which is only needed if you want to reach a backend a second time (perhaps with a slightly modified query, or selecting a different backend). I think what you need is a way to generate a custom response from a `DNSResponse`, so some equivalent of `DNSQuestion:spoof()` or `SpoofAction` that will work on a response. It doesn't seem to be implemented right now but at the moment I don't see any reason why we would not want to implement that (we need to be careful about not caching the custom response, though).

**Reply by KagurazakaNyaa** [reply]
> Right, then I'm not sure you actually need to restart a query, which is only needed if you want to reach a backend a second time (perhaps with a slightly modified query, or selecting a different backend). I think what you need is a way to generate a custom response from a `DNSResponse`, so some equivalent of `DNSQuestion:spoof()` or `SpoofAction` that will work on a response. It doesn't seem to be implemented right now but at the moment I don't see any reason why we would not want to implement that (we need to be careful about not caching the custom response, though).

Indeed, I tried to use the tag+restart method because https://dnsdist.org/reference/dq.html#DNSResponse mentioned that spoof cannot be used for response. If this feature is implemented directly, restart is not needed.

**Reply by rgacogne** [reply]
In the meantime, one option might be to add a packet cache to your pool, and call `SpoofAction` from the "cache-miss" chain (so added via `addCacheMissAction`, not that this requires running a master build of dnsdist since this directive/rules chain was added after 1.9.x). I did not test it but in theory it should work.

**Reply by KagurazakaNyaa** [reply]
> In the meantime, one option might be to add a packet cache to your pool, and call `SpoofAction` from the "cache-miss" chain (so added via `addCacheMissAction`, not that this requires running a master build of dnsdist since this directive/rules chain was added after 1.9.x). I did not test it but in theory it should work.

It can be noted from the document that the Action type of `addCacheMissAction` is `DNSDistRuleAction` instead of `DNSDistResponseRuleAction`, so the matching Action type is `DNSAction` instead of `DNSResponseAction`. The parameter passed in to this Action seems to be `dq` instead of `dr`, so the returned content cannot be parsed correctly.
Maybe I should insert the rules into `addCacheHitResponseAction` or `addCacheInsertedResponseAction`?

**Reply by rgacogne** [reply]
The rules chain related to `addCacheMissAction` is a question rules chain, not a a response one, yes. That's the whole point since you want to get a `DNSQuestion` object to be able to use `DNSQuestion:spoof()` from a custom Lua rule, or `SpoofAction`. If you try to use a response rules chain you will gain nothing compared to `addResponseAction`.

**Reply by rgacogne** [reply, marked as answer]
So, what I'm suggesting is:
- do the parsing in a response rule chain, like you are doing today. Restart your query after setting a tag like you are currently doing, ensuring there is a packet cache associated to the pool you are routing the restarted query to ;
- intercept the restarted query via `addCacheMissAction` instead of `addAction`, matching the tag as needed, and there you can call `SpoofAction`.

**Reply by KagurazakaNyaa** [reply]
> So, what I'm suggesting is:
> 
>     * do the parsing in a response rule chain, like you are doing today. Restart your query after setting a tag like you are currently doing, ensuring there is a packet cache associated to the pool you are routing the restarted query to ;
> 
>     * intercept the restarted query via `addCacheMissAction` instead of `addAction`, matching the tag as needed, and there you can call `SpoofAction`.

Tested on `powerdns/dnsdist-master:latest`, this method is effective
